### PR TITLE
feat(rds): support certificate autority certificate

### LINKED ISSUE #26865: (RDS): Support RDS certificate authority
### Latest Update from CDK Team:

Still missing support for reader/writers on cluster. See:
https://github.com/aws/aws-cdk/issues/26865#issuecomment-1716153176

### Describe the feature

Since cloudformation supports changing the rds certificate authority CDK should do so as well. aws-cloudformation/cloudformation-coverage-roadmap#211.

### Use Case

My current CA is rds-ca-2019 and I need to change it. All our AWS resources are scripted as IaC. Therfore, I do not want to change that via the AWS Console but use CDK instead.

### Proposed Solution

Add a property supporting the CA

### Other Information

To contribute a PR for this I probably would need some guidance since I this would be my first contribution to the CDK repository

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.91.0

### Environment details (OS name and version, etc.)

macOS 13.5.1

### PR DESCRIPTION
**Pull request by lpizzinidev**
Exposes the `caCertificate` property for an RDS instance to allow specifying a custom CA identifier using the `CaCertificate` class.

Usage:
```
new DatabaseInstance(stack, 'Instance', {
      engine:DatabaseInstanceEngine.mysql({ version: rds.MysqlEngineVersion.VER_8_0_30 }),
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.BURSTABLE3, ec2.InstanceSize.SMALL),
      vpc,
      caCertificate: CaCertificate.RDS_CA_RSA2048_G1,
});
```

Closes #26865.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-rds/lib/instance.ts [hunk 1]
```diff
@@ -713,6 +738,20 @@ export interface DatabaseInstanceNewProps {
...
+   * Specifying or updating this property triggers a reboot.
+   *
+   * For RDS DB engines:
+   * @see https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.SSL-certificate-rotation.html
+   * For Aurora DB engines:
+   * @see https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/UsingWithRDS.SSL-certificate-rotation.html
+   *
+   * @default - rds-ca-2019
+   */
+  readonly caCertificateIdentifier?: CertificateIdentifier;
```
Comment (mrgrain) [inline packages/aws-cdk-lib/aws-rds/lib/instance.ts:754]:
```suggestion
  readonly caCertificate?: CertificateIdentifier;
```

Can you drop the `Identifier` everywhere please. It's not really relevant to the user intent.
Also in the docstring, e.g. `The the CA certificate used for this DB instance.`

File: packages/aws-cdk-lib/aws-rds/lib/instance.ts [hunk 2]
```diff
@@ -347,6 +347,31 @@ export enum NetworkType {
   DUAL = 'DUAL'
 }
 
+/**
+ * The CA identifier of the DB instance
+ */
+export enum CertificateIdentifier {
```
Comment (mrgrain) [inline packages/aws-cdk-lib/aws-rds/lib/instance.ts:353]:
```suggestion
export enum CaCertificate {
```

Comment (mrgrain) [inline packages/aws-cdk-lib/aws-rds/lib/instance.ts:353]: Can you turn the enum into an enum like class instead, with the ability to create a CaCertificate from a custom string: `new CaCertificate('future-identifier')`

```ts
class CaCertificate {
  static public RDS_CA_2019() { return new CaCertificate('rds-ca-2019') }
  public constructor(identifier: string) { }
}
```

**Review by mrgrain — CHANGES_REQUESTED** [review]
Nice one, thank you. Got just a couple requests.

**Comment by robertd** [comment]
@mrgrain @lpizzinidev I'm worming up to enum-like class approach. Would this approach be better than getters & setters? In the code we'd use `CaCertificate.RDS_CA_2019` compared to `CaCertificate.rdsCaRsa2019()` which feels more like invoking an enum. Thoughts?

```ts
export class CaCertificate {
  /**
   * rds-ca-2019 certificate authority
   */
  public static readonly RDS_CA_2019 = CaCertificate.of('rds-ca-2019');

  /**
   * rds-ca-ecc384-g1 certificate authority
   */
  public static readonly RDS_CA_ECC384_G1 = CaCertificate.of('rds-ca-ecc384-g1');

  /**
   * rds-ca-rsa2048-g1 certificate authority
   */
  public static readonly RDS_CA_RDS2048_G1 = CaCertificate.of('rds-ca-rsa2048-g1');

  /**
   * rds-ca-rsa4096-g1 certificate authority
   */
  public static readonly RDS_CA_RDS4096_G1 = CaCertificate.of('rds-ca-rsa4096-g1');

  /**
   * Custom certificate ID
   * @param version custom certificate authority ID
   */
  public static of(version: string) { return new CaCertificate(version); }

  /**
   *
   * @param version certificate authority ID
   */
  private constructor(public readonly version: string) { }
}
```

**Comment by mrgrain** [comment]
@robertd Thanks for the review! Yes this is exactly what I had in mind. Apologies for not conveying this in my review.

I think in general it's the better pattern because it enables us to easily extend features in future. E.g. what if (theoretically)  you could upload your own CA Certificate in future. It's more strongly typed and unlike enums can be extended on by userland code. In the wider TS community, there's some talk at the moment about how enums are bad in general.

So yes, we probably should use enum-like classes everywhere instead.

In practice, it might more relevant for some situations than others. I picked it up here, because historically with certificates there is ALWAYS going to be a later one.

**Comment by robertd** [comment]
@mrgrain Yep, makes total sense in this use case.  If @lpizzinidev can change enum like class to use `CaCertificate.RDS_CA_2019` instead of `CaCertificate.rdsCaRsa2019()` (and update stuff in other places accordingly) then I think we’re good to ship this. 👍 


> ```ts
> export class CaCertificate {
>   /**
>    * rds-ca-2019 certificate authority
>    */
>   public static readonly RDS_CA_2019 = CaCertificate.of('rds-ca-2019');
> 
>   /**
>    * rds-ca-ecc384-g1 certificate authority
>    */
>   public static readonly RDS_CA_ECC384_G1 = CaCertificate.of('rds-ca-ecc384-g1');
> 
>   /**
>    * rds-ca-rsa2048-g1 certificate authority
>    */
>   public static readonly RDS_CA_RDS2048_G1 = CaCertificate.of('rds-ca-rsa2048-g1');
> 
>   /**
>    * rds-ca-rsa4096-g1 certificate authority
>    */
>   public static readonly RDS_CA_RDS4096_G1 = CaCertificate.of('rds-ca-rsa4096-g1');
> 
>   /**
>    * Custom certificate identifier
>    * @param custom certificate authority identifier
>    */
>   public static of(identifier: string) { return new CaCertificate(identifier); }
> 
>   /**
>    *
>    * @param certificate authority identifier
>    */
>   private constructor(public readonly identifier: string) { }
> }
> ```

**Comment by robertd** [comment]
@lpizzinidev is on the top of the game with this PR with updating it to the latest changes we discussed. LGTM. 👍

File: packages/aws-cdk-lib/aws-rds/lib/instance.ts [hunk 3]
```diff
@@ -347,6 +347,45 @@ export enum NetworkType {
   DUAL = 'DUAL'
 }
 
+/**
+ * The CA certificate used for this DB instance.
```
Comment (robertd) [inline packages/aws-cdk-lib/aws-rds/lib/instance.ts:351]: One minor nitpick… add `@see: https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.SSL.html` in jsdocs section.

**Comment by lpizzinidev** [comment]
@robertd Thanks for the feedback.
I missed the blueprint used for enum-like classes in the rest of the project.
Now the code is much cleaner 👍

File: packages/aws-cdk-lib/aws-rds/lib/instance.ts [hunk 4]
```diff
@@ -713,6 +754,20 @@ export interface DatabaseInstanceNewProps {
...
+   * The identifier of the CA certificate for this DB instance.
+   *
+   * Specifying or updating this property triggers a reboot.
+   *
+   * For RDS DB engines:
+   * @see https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.SSL-certificate-rotation.html
+   * For Aurora DB engines:
+   * @see https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/UsingWithRDS.SSL-certificate-rotation.html
+   *
+   * @default - CaCertificate.RDS_CA_2019
```
Comment (mrgrain) [inline packages/aws-cdk-lib/aws-rds/lib/instance.ts:768]: This default is unexpected and seems like a potentially breaking change to me. Can you explain or remove it?

File: packages/aws-cdk-lib/aws-rds/lib/instance.ts [hunk 5]
```diff
@@ -865,6 +920,7 @@ abstract class DatabaseInstanceNew extends DatabaseInstanceBase implements IData
       domain: this.domainId,
       domainIamRoleName: this.domainRole?.roleName,
       networkType: props.networkType,
+      caCertificateIdentifier: props.caCertificate ? props.caCertificate.identifier : CaCertificate.RDS_CA_2019.identifier,
```
Comment (mrgrain) [inline packages/aws-cdk-lib/aws-rds/lib/instance.ts:923]: Default question, see above.

```suggestion
      caCertificateIdentifier: props.caCertificate ? props.caCertificate.identifier : undefined,
```

Btw, we also could implement a `toString()` method on `CaCertificate`. This is works as well though.

**Review by mrgrain — CHANGES_REQUESTED** [review]
One last question regarding the default.

File: packages/aws-cdk-lib/aws-rds/lib/instance.ts (hunk 4 above)
Comment (lpizzinidev) [reply packages/aws-cdk-lib/aws-rds/lib/instance.ts:768]: Trying to create an instance from the console, it suggests `rds-ca-2019` as default (this is the reason for the `@default`).

Screenshot 2023-08-28 at 15 53 43

Anyway, looking at the CFN [documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbinstance.html#cfn-rds-dbinstance-cacertificateidentifier), it does not specify any default, so I guess I should remove it.

Comment (mrgrain) [reply packages/aws-cdk-lib/aws-rds/lib/instance.ts:768]: I see. It probably doesn't change much, but I think we want to end up in this case:
"If you don't select a certificate authority, RDS chooses one for you."

```suggestion
   * @default - Let RDS chooses a certificate authority for you
```

**Review by mrgrain — APPROVED** [review]

**Review by robertd — APPROVED** [review]
LGTM. Many thanks @lpizzinidev 🚀

**Review by mrgrain — APPROVED** [review]

**Comment by StefanPrintezis** [comment]
static .of() usually refers to a known given, like an enum. Why would the constructor be private and then publish the of( method to be  a constructor wrapper, it's a bit inverse right now. If you want a custom CA i'd think is fine `new PrivateCa('')` , like InstanceType.

I think I'd still prefer to go with the enums for consistency. Feels more light weight than having the instances be instantiated in properties of a static class.


```
export declare enum CertificateIdentifier {
    /**
     * rds-ca-2019 certificate authority
     */
    RDS_CA_2019 = 'rds-ca-2019',
    /**
     * rds-ca-ecc384-g1 certificate authority
     */
    RDS_CA_ECC384_G1 = 'rds-ca-ecc384-g1',
    /**
     * rds-ca-rsa2048-g1 certificate authority
     */
    RDS_CA_RDS2048_G1 = 'rds-ca-rsa2048-g1',
    /**
     * rds-ca-rsa4096-g1 certificate authority
     */
    RDS_CA_RDS4096_G1 = 'rds-ca-rsa4096-g1',
}
/**
 * The CA certificate used for this DB instance.
 *
 * This class takes a literal string, good if you already
 * know the identifier of the type you want.
 */
export declare class CaCertificate {

    static of(identifier: CertificateIdentifier): CaCertificate {
        return new CaCertificate(identifier);
    }
    constructor(private readonly identifier: string) {};

    /**
     * Return the instance type as a string.
     */
    toString(): string {
        return this.identifier;
    }
}

```

**Comment by mrgrain** [comment]
> static .of() usually refers to a known given, like an enum. Why would the constructor be private and then publish the of( method to be a constructor wrapper, it's a bit inverse right now. If you want a custom CA i'd think: `new PrivateCa('custom-ca')`
> 
> I think I'd still prefer to go with the enums for consistency. Feels more light weight than having the instances be instantiated in properties of a static class.

But using it would still be typed to `CaCertificate`, so as a user you will have to do:

```ts
new CaCertificate(CertificateIdentifier. RDS_CA_2019)
```

and instantiate anyway.

`static of(identifier: string)` is the pattern we use right now everywhere else.

**Comment by mrgrain** [comment]
@mergifyio requeue

**Comment by baguilarq** [comment]
do we need another implementation to support this for `DatabaseCluster`, isn't it?

**Pull request merged**